### DISCUSSION TITLE
Please tell me how can i build openssl.cmake that build for opeen ssl

### DISCUSSION THREAD

**Discussion by josephjw** [schwabe/ics-openvpn#1747] [Q&A]
i need to know how you generated the openssl.cmake file. it will help me to link oqs . (atleast for 1.1.1u)

Please help later using that i need to do integrate oqs provider also

**Comment by schwabe** [comment]
The file is not generated. The file was just written to build OpenSSL. Sure the defines were taken/adopted from the old Android build system with the mk/what OpenSSL generates itself but ultimatively the file is not generated. To include the qos provider you might just build that as static library and then link that to OpenVPN and call the add internal provider method like we already do with xkey

**Reply by josephjw** [reply]
I need to link the crypto and liboqs as mentioned below so that i can access kyber1024 for openssl handshake.

i couldnt link with static library it shows error "libcrypto" not build by this project. so if i get openssl.cmake file for openssl1.1.1u it will be great.   

here i m getting kybernot found    " int nid = OBJ_sn2nid(sn);" 

 I linked crypto oqs with main project in this case i can access kyber from only my own file 
like targetling(project crypto oqs )

but i want pass kyber for openssl function  like below 
My native.cpp file below 

```
  // const char *sn = "kyber512";
    const char *sn = "kyber1024";
    // Convert short name to NID
    int nid = OBJ_sn2nid(sn);
    if (nid == NID_undef)
    {
        syslog(LOG_CRIT, "[Cl-con] Unknown object short name: %s\n", sn);
        // return 1;
        exit(1);
    }
    
 ```
    *****************************************************************************************************
    
    
```
            crypto/ec/asm/x25519-x86_64.S
            crypto/x86_64cpuid.S
            )

    list(REMOVE_ITEM crypto_srcs
            crypto/aes/aes_core.c
            crypto/bn/bn_asm.c
            crypto/mem_clr.c
            crypto/rc4/rc4_enc.c
            crypto/rc4/rc4_skey.c
            )




else()
    message( FATAL_ERROR "Unknown arch ${ANDROID_ABI} for source files" )
endif()

PREPEND(crypto_srcs_with_path ${OPENSSL_PATH} ${crypto_srcs})
add_library(crypto ${SSLLIBTYPE} ${crypto_srcs_with_path} )
target_link_libraries(crypto
                        oqs
                       )
                        
```                        
                        
 another hand if i get openss.cmake file i can also do like above

**Reply by schwabe** [reply]
I would try to avoid using OBJ_sn2nid in OpenSSL 3.0, especially not for external ciphers provided by provider as it relies on static allocation of integers.

**Comment by josephjw** [comment]
Atleast may i get openssl.cmake file for openssl1.1.1u 

i need to call quantum algo inside the openssl. 

const char *sn = "kyber1024";
// Convert short name to NID
int nid = OBJ_sn2nid(sn);

so give me if you have openssl.cmake or tell how can i get it. i dont same thing for ubuntu and windows but here i m facing issue in linking

**Reply by schwabe** [reply]
Check the `oldopenssl` branch of the repoif there is a version of openssl.cmake that works with OpenSSL 1.1.1u.

But if there is not one, you will have to write one for your own. I don't think you can expect to write one for you since I am not sure what my motivation there would be.

**Comment by josephjw** [comment]
first things is openssl1.1.1e for that openssl.cmake 

below these all file are not found  for openssl 1.1.1u when i m suing same openssl.cmkae file
elseif(${ANDROID_ABI} STREQUAL "arm64-v8a")
    set(crypto_srcs ${crypto_srcs}
#            crypto/aes/asm/aesv8-armx-64.S
#            crypto/aes/asm/vpaes-armv8.S
            crypto/arm64cpuid.S
#            crypto/bn/asm/armv8-mont.S
#            crypto/ec/asm/ecp_nistz256-armv8.S
#            crypto/modes/asm/ghashv8-armx-64.S
            crypto/poly1305/asm/poly1305-armv8.pl
            crypto/sha/asm/sha1-armv8.pl
            crypto/sha/asm/sha256-armv8.pl)
            
            
on the other hand,  i used openssl 1.1.1e to test with my server which has openssl 3.2.1. from mobile i got error 👎🏻 

[Cl-con] Unknown object short name: kyber1024

at the line of 

// const char *sn = "kyber512";
    const char *sn = "kyber1024";
    // Convert short name to NID
    int nid = OBJ_sn2nid(sn);
    if (nid == NID_undef)
    {
        syslog(LOG_CRIT, "[Cl-con] Unknown object short name: %s\n", sn);
        // return 1;
        exit(1);
    }
    
    
    
    how can i link oqs with  i tried like this below 
<img width="692" alt="Screenshot 2024-08-01 at 4 28 42 PM" src="https://github.com/user-attachments/assets/3a698d1a-71d1-43bc-9f8a-8fb850d858eb">


should i need oqsprov , i hope it s from 3.0 onwards 


i need to make enable oqs enabled in 1.1.1u or 3.2.1 based openssl,  
<img width="964" alt="Screenshot 2024-08-01 at 4 30 46 PM" src="https://github.com/user-attachments/assets/21800b91-2d90-44e8-b782-fbc47645e98c">

above like this added oqs , oqsprov and  i tried still it s saying kyber1024 is Unknown object.


please Do needful. 

thanks

**Reply by schwabe** [reply]
OpenSSL 1.1.1 is out of support. I am not spending more time with it.

**Comment by josephjw** [comment]
My requirement is simple need to talk to server it has 3.2.1 version of openssl.

 from mobile client i need to talk to server and get cert files.

server enables quantum algo , so i need to select kyber1024.

**Reply by schwabe** [reply]
Sure. But adding qos provider seems to be more than a just 5 minute job. So either I need to have a strong motivation to do invest significant time into helping you on this endavor or you need to figure it out by yourself. If you are working for a company you/your company can contact me for contracting doing this work. Or need a strong motivation/background/use case for an OSS use case. Otherwise I am not sure why I should spend my time developing something for you that I am currently not interested in.

**Reply by josephjw** [reply]
Please give me your contact information including email and phone number. I need to schedule the call.

email: infotostalin@gmail.com

**Reply by josephjw** [reply]
Let's go head quickly and tell me when we can connect for it.

to proceed further
